src/external_acl.cc

```diff
@@ -968,28 +968,32 @@ externalAclHandleReply(void *data, const Helper::Reply &reply)
 
     entryData.notes.append(&reply.notes);
 
-    const char *label = reply.notes.findFirst("tag");
-    if (label != nullptr && *label != '\0')
-        entryData.tag = label;
+    NotePairs replyNotes;
+    replyNotes.append(&reply.notes);
 
-    label = reply.notes.findFirst("message");
-    if (label != nullptr && *label != '\0')
-        entryData.message = label;
+    auto label = replyNotes.findFirstAndRemove("tag");
+    if (label.length())
+        entryData.tag.assign(label.rawContent(), label.length());
 
```

Comment:
This append-all-then-extract-some-then-check-leftovers design adds quite a bit of performance overhead. I hope we do not have to do that. The code also continues to look ugly.

Would it be possible to refactor this to arrive at something like the _sketch_ below?

```C++
entryData.notes.importIf(reply.notes, [&](const auto &note) {
    if (note.name == "tag") {
        entryData.tag = note.value;
        return false;
    }
    if (note.name == "message") {
        entryData.message = note.value;
        return false;
    }
    ...
    return CheckCustomAnnotation(note, "external_acl");
});
```

Reply:
I tried to implement this approach in a separate [PR182](https://github.com/measurement-factory/squid/pull/182/files), please have a look. I covered some of the existing contexts - I plan to adjust the other ones as soon as we agree upon the overall design.

Reply:
Thank you. I am reviewing/working on #182 now.

Reply:
After my importIf() idea [failed](https://github.com/measurement-factory/squid/pull/182#pullrequestreview-1231071067) in #182, I attacked the problem from a different angle, resulting in #183.

@eduard-bagdasaryan, please take a look at that draft PR (no in-depth/detailed review is needed at this point -- just check for any undisclosed high-level problems). #183 is not without its own problems (disclosed in the PR description), but at least it does not make the caller code look worse. In fact, nearly all of the official callers can be preserved "as is" in that approach if we decide not to rename existing methods! I may post another draft PR with a completely different (and much simpler) solution, but I may need another 24 hours for that.

Reply:
> I may post another draft PR with a completely different (and much simpler) solution

Done at #184.